README.rst
.. -*- coding: utf-8 -*-
.. :Project:   metapensiero.asyncio.transaction -- Handle coroutines from synchronous functions or methods (like special methods)
.. :Created:   dom 09 ago 2015 12:57:35 CEST
.. :Author:    Alberto Berti <alberto@metapensiero.it>
.. :License:   GNU General Public License version 3 or later
.. :Copyright: Copyright (C) 2015 Alberto Berti
..

==================================
 metapensiero.asyncio.transaction
==================================

 :author: Alberto Berti
 :contact: alberto@metapensiero.it
 :license: GNU General Public License version 3 or later

Handle coroutines from synchronous functions or methods (like special methods)
------------------------------------------------------------------------------

Goal
~~~~

This package helps handling such cases when there is a side effect
that is coded as a coroutine and the code block that needs it cannot
be run as a coroutine.

The case is e. g. when a coroutine is called from an external package
or a Python special method.

This package will give you the tools to ensure that even if the
computation is not immediate, the execution of the coroutine(s) is
guaranteed to happen in the context of the coroutine that is calling
the special method.

A package that works in tandem with this is `metapensiero.signal`__

__ https://pypi.python.org/pypi/metapensiero.signal

Installation
~~~~~~~~~~~~

To install the package execute the following command::

  $ pip install metapensiero.asyncio.transaction

Usage
~~~~~

Given a scenario where a coroutine is called from the ``__setattr___``
method, this is how to deal with the situation:

.. code:: python

  import asyncio
  from metapensiero.async import transaction

  @asyncio.coroutine
  def publish(value):
     # do something async

  class Example:

      def __setattr__(self, name, value):
          trans = transaction.get()
          trans.add(publish(value))
          super().__setattr__(name, value)

  @asyncio.coroutine
  def external_coro():
      inst = Example()
      trans = transaction.begin()
      inst.foo = 'bar'
      yield from trans.end()

In python 3.5, the ``external_coro`` can be written as:

.. code:: python

  async def external_coro():
      inst = Example()
      async with transaction.begin():
          inst.foo = 'bar'

So by taking advantage of the new ``__aenter__`` and ``__aexit__``
methods and awaitable classes.

The coroutines will be scheduled in the order they have been created.

At a certain point, you may want  to ensure that all the remaining
coroutines are executed you may use the coroutine
``transaction.wait_all()``, doing so will end all the remaining open
transactions.

When your code add a coro to the transaction it can also pass a
callback as the ``cback`` keyword parameter, so it will be called when
the stashed coro completes (or is cancelled). This way you can use the
return value of the coro like:

.. code:: python

  import asyncio
  from metapensiero.asyncio import transaction

  results = []

  @asyncio.coroutine
  def stashed_coro():
      nonlocal results
      results.append('called stashed_coro')
      return 'result from stashed coro'

  class A:

      def __init__(self):
          tran = transaction.get()
          c = stashed_coro()
          tran.add(c, cback=self._init)

      def _init(self, stashed_task):
          nonlocal results
          results.append(stashed_task.result())

  @asyncio.coroutine
  def external_coro():
      tran = transaction.begin()
      # in py3.5
      # async with tran:
      #     a = A()
      a = A()
      yield from tran.end()

  yield from asyncio.gather(
      external_coro()
  )

  assert len(results) == 2
  assert results == ['called stashed_coro', 'result from stashed coro']


Testing
~~~~~~~

To run the tests you should run the following at the package root::

  python setup.py test


Build status
~~~~~~~~~~~~

.. image:: https://travis-ci.org/azazel75/metapensiero.asyncio.transaction.svg?branch=master
    :target: https://travis-ci.org/azazel75/metapensiero.asyncio.transaction


CHANGES.rst
.. -*- coding: utf-8 -*-

Changes
-------

0.1 (2015-12-15)
~~~~~~~~~~~~~~~~

- Adaptation from the rocky sources.
- First draft documentation.

0.2 (2015-12-16)
~~~~~~~~~~~~~~~~

- Document some details about testing.
- Add some checks on end().

0.3 (2015-12-26)
~~~~~~~~~~~~~~~~

- Fix packaging.

0.4 (2015-12-27)
~~~~~~~~~~~~~~~~

- Documentation fixes.

0.5 (2016-01-22)
~~~~~~~~~~~~~~~~

- Fix behavior when transaction.begin() is called from code not
  already running in a task.

0.6 (2016-01-27)
~~~~~~~~~~~~~~~~

- support a callback function when adding coros to the transaction.

0.7 (2016-02-18)
~~~~~~~~~~~~~~~~

- Add a wait() method to suspend the current task while the coros complete
  without closing the current transaction.


src/metapensiero/asyncio/transaction/tests/test_transaction.py
# -*- coding: utf-8 -*-
# :Project:  metapensiero.asyncio.transaction -- tests
# :Created:    mar 15 dic 2015 15:16:56 CET
# :Author:    Alberto Berti <alberto@metapensiero.it>
# :License:   GNU General Public License version 3 or later
#

import asyncio

import pytest

from metapensiero.asyncio import transaction

@pytest.mark.asyncio
@asyncio.coroutine
def test_transaction_per_task(event_loop):

    tasks_ids = set()
    results = []

    @asyncio.coroutine
    def stashed_coro():
        nonlocal results
        results.append('called stashed_coro')

    def non_coro_func():
        tran = transaction.get(loop=event_loop)
        c = stashed_coro()
        tran.add(c)

    @asyncio.coroutine
    def external_coro():
        nonlocal tasks_ids
        task = asyncio.Task.current_task(loop=event_loop)
        tasks_ids.add(id(task))
        tran = transaction.begin(loop=event_loop)
        # in py3.5
        # async with tran:
        #     non_coro_func()
        non_coro_func()
        yield from tran.end()


    yield from asyncio.gather(
        external_coro(),
        external_coro(),
        loop=event_loop
    )

    assert len(tasks_ids) == 2
    assert len(results) == 2
    assert results == ['called stashed_coro', 'called stashed_coro']

@pytest.mark.asyncio
@asyncio.coroutine
def test_non_closed_transaction(event_loop):

    tasks_ids = set()
    results = []

    event_loop.set_debug(True)

    @asyncio.coroutine
    def stashed_coro():
        nonlocal results
        results.append('called stashed_coro')

    def non_coro_func():
        tran = transaction.get(loop=event_loop)
        c = stashed_coro()
        tran.add(c)

    @asyncio.coroutine
    def external_coro():
        nonlocal tasks_ids
        task = asyncio.Task.current_task(loop=event_loop)
        tasks_ids.add(id(task))
        tran = transaction.begin(loop=event_loop)
        # in py3.5
        # async with tran:
        #     non_coro_func()
        non_coro_func()


    yield from external_coro()
    done, pending = yield from transaction.wait_all()
    # the raise from the callback gets sucked up, this is the only
    # crumb left
    assert len(done) == 1
    assert len(pending) == 0

def test_calling_from_non_task():

    tasks_ids = set()
    results = []

    @asyncio.coroutine
    def stashed_coro():
        nonlocal results
        results.append('called stashed_coro')

    def non_coro_func():
        tran = transaction.get()
        c = stashed_coro()
        tran.add(c)

    # create a new event loop because i get a closed one with
    # get_event_loop(), probably a conflict with pytest-asyncio
    loop = asyncio.new_event_loop()
    tran = transaction.begin(loop=loop)
    non_coro_func()
    assert len(results) == 0
    loop.run_until_complete(tran.end())
    assert len(results) == 1

@pytest.mark.asyncio
@asyncio.coroutine
def test_transaction_per_task_with_cback(event_loop):

    results = []

    @asyncio.coroutine
    def stashed_coro():
        nonlocal results
        results.append('called stashed_coro')
        return 'result from stashed coro'

    def non_coro_func():
        tran = transaction.get(loop=event_loop)
        c = stashed_coro()
        tran.add(c, cback=_cback)

    def _cback(stashed_task):
        results.append(stashed_task.result())

    @asyncio.coroutine
    def external_coro():
        tran = transaction.begin(loop=event_loop)
        # in py3.5
        # async with tran:
        #     non_coro_func()
        non_coro_func()
        yield from tran.end()

    yield from asyncio.gather(
        external_coro(),
        loop=event_loop
    )

    assert len(results) == 2
    assert results == ['called stashed_coro', 'result from stashed coro']

@pytest.mark.asyncio
@asyncio.coroutine
def test_transaction_per_task_with_cback2(event_loop):

    results = []

    @asyncio.coroutine
    def stashed_coro():
        nonlocal results
        results.append('called stashed_coro')
        return 'result from stashed coro'

    class A:

        def __init__(self):
            tran = transaction.get(loop=event_loop)
            c = stashed_coro()
            tran.add(c, cback=self._init)

        def _init(self, stashed_task):
            nonlocal results
            results.append(stashed_task.result())

    @asyncio.coroutine
    def external_coro():
        tran = transaction.begin(loop=event_loop)
        # in py3.5
        # async with tran:
        #     a = A()
        a = A()
        yield from tran.end()

    yield from asyncio.gather(
        external_coro(),
        loop=event_loop
    )

    assert len(results) == 2
    assert results == ['called stashed_coro', 'result from stashed coro']


src/metapensiero/asyncio/transaction/__init__.py
# -*- coding: utf-8 -*-
# :Project:   metapensiero.asyncio.transaction -- Handle coroutines from synchronous functions or methods (like special methods)
# :Created:   dom 09 ago 2015 12:57:35 CEST
# :Author:    Alberto Berti <alberto@metapensiero.it>
# :License:   GNU General Public License version 3 or later
# :Copyright: Copyright (C) 2015 Alberto Berti
#

import asyncio
import functools
import inspect
import logging
import sys
import weakref

PY34 = sys.version_info >= (3, 4)
PY35 = sys.version_info >= (3, 5)

logger = logging.getLogger(__name__)

TRANSACTIONS = {}

RELAXED_TRANSACTION_END = False

_nodefault = object()

__all__ = ('Transaction', 'get', 'begin', 'end', 'wait_all', 'RELAXED_TRANSACTION_END')

try:
    # travis tests compatibility?!
    from asyncio import ensure_future
except ImportError:
    ensure_future = asyncio.async

class TransactionError(Exception):
    pass


class Transaction:
    """A mechanism to store coroutines and consume them later.
    """

    @classmethod
    def get(cls, default=_nodefault, loop=None, registry=None, task=None):
        """Get the ongoing transaction for the current task. if a current
        transaction is missing either raises an exception or returns
        the passed-in 'default'.
        """
        global TRANSACTIONS
        task = task or asyncio.Task.current_task(loop=loop)
        registry = registry or TRANSACTIONS
        task_id = hash(task)
        trans_list = registry.get(task_id)
        if trans_list:
            result = trans_list[-1]
        else:
            if default is _nodefault:
                raise TransactionError("There's no transaction"
                                       " begun for task %s" % task_id)
            else:
                result = default
        return result

    @classmethod
    def begin(cls, loop=None, registry=None, task=None):
        """Begin a new transaction"""
        global TRANSACTIONS
        task = task or asyncio.Task.current_task(loop=loop)
        registry = registry or TRANSACTIONS
        task_id = hash(task)
        trans_list = registry.get(task_id)
        if not trans_list:
            registry[task_id] = trans_list = []
        trans = cls((task_id, len(trans_list)), loop, registry)
        trans_list.append(trans)
        if task:
            # task my be None because the code isn't scheduled by asyncio
            task.add_done_callback(functools.partial(cls._owner_task_finalization_cb,
                                                 weakref.ref(trans)))
        return trans

    @classmethod
    def _owner_task_finalization_cb(cls, trans_ref, task):
        """Warn about non-automatic one left open.
        """
        trans = trans_ref()
        if trans and trans.open:
            msg = ("A transaction has not been closed: %r", trans)
            if RELAXED_TRANSACTION_END:
                logger.warning(*msg)
            else:
                logger.error(*msg)
                raise TransactionError(msg[0] % msg[1])

    def __init__(self, trans_id, loop=None, registry=None):
        self.registry = registry or TRANSACTIONS
        self.coros = []
        self.loop = loop or asyncio.get_event_loop()
        self.id = trans_id
        self.open = True
        logger.debug('Beginning transaction: %r', self)

    def add(self, *coros, cback=None):
        """Add a coroutine or awaitable to the set managed by this
        transaction.
        """
        for coro in coros:
            if PY35:
                assert inspect.isawaitable(coro)
            if asyncio.iscoroutine(coro):
                task = ensure_future(coro, loop=self.loop)
            else:
                task = coro
            if task not in self.coros:
                if isinstance(task, asyncio.Future):
                    task.add_done_callback(self._task_remove_cb)
                    if cback:
                        task.add_done_callback(cback)
                self.coros.append(task)
                logger.debug("Added task %r to trans %r", task, self)

    @asyncio.coroutine
    def end(self):
        """Close an ongoing transaction."""
        result = yield from self.wait()
        logger.debug('Ending transaction: %r', self)
        self.open = False
        self.remove(self)
        del self.registry
        del self.coros
        del self.loop
        return result

    @classmethod
    def remove(cls, trans):
        """Remove a transaction from its registry."""
        registry = trans.registry
        trans_list = registry[trans.id[0]]
        assert len(trans_list) > 0
        top_trans = trans_list.pop()
        assert trans is top_trans
        if len(trans_list) == 0:
            del registry[trans.id[0]]

    @asyncio.coroutine
    def __aenter__(self):
        return self

    @asyncio.coroutine
    def __aexit__(self, exc_type, exc, tb):
        if not exc:
            yield from self.end()
        else:
            logger.debug('An error happened on context exit:',
                         exc_info=(exc_type, exc, tb))
        return False

    def __repr__(self):
        return '<%s id: %s number of items: %d state: %s>' % \
            (self.__class__.__name__, self.id, len(self.coros),
             'open' if self.open else 'closed')

    @staticmethod
    @asyncio.coroutine
    def wait_all(timeout=None, loop=None, registry=None):
        """Return a future that will be complete when the pending coros of
        the transactions will complete, effectively ending all of them.
        """
        # TODO: take loop into account
        global TRANSACTIONS
        registry = registry or TRANSACTIONS
        loop = loop or asyncio.get_event_loop()

        # collect pending transactions
        coros = set()
        for task_id, transactions in registry.items():
            for trans in transactions:
                coros.add(trans.end())
        if coros:
            result = asyncio.wait(coros, loop=loop, timeout=timeout)
        else:
            result = None
        return result

    def _task_remove_cb(self, task):
        """Remove a scheduled coroutine from the set of this transaction."""
        self.coros.remove(task)
        logger.debug("Removed task %r from trans %r", task, self)

    @asyncio.coroutine
    def wait(self):
        """Wait for this coros to complete, expunge them but not close this
        transaction. Useful when there is one 'global' transaction per taks.
        """
        if not self.open:
            raise TransactionError("This transaction is closed already")
        logger.debug('Waiting for this transaction coros to complete: %r', self)
        result = yield from asyncio.gather(*self.coros, loop=self.loop)
        self.coros.clear()
        return result

get = Transaction.get
begin = Transaction.begin


@asyncio.coroutine
def end(loop=None, registry=None, task=None):
    """End the current defined transaction."""
    trans = get(None, loop, registry, task)
    return (yield from trans.end())

wait_all = Transaction.wait_all

@asyncio.coroutine
def wait(loop=None, registry=None, task=None):
    """Wait for the current defined transaction's coroutines to complete."""
    trans = get(None, loop, registry, task)
    return (yield from trans.wait())
